cmd/keymasterd/2fa_okta.go

```diff
@@ -0,0 +1,171 @@
+	if !valid {
+		logger.Printf("Invalid OTP value login for %s", authUser)
+		// TODO if client is html then do a redirect back to vipLoginPage
+		state.writeFailureResponse(w, r, http.StatusUnauthorized, "")
+		return
+
+	}
+	// OTP check was  successful
+	logger.Debugf(1, "Successful Okta OTP auth for user: %s", authUser)
+
+	// TODO ADD okta events to eventmond
+	//eventNotifier.PublishVIPAuthEvent(eventmon.VIPAuthTypeOTP, authUser)
+
+	_, err = state.updateAuthCookieAuthlevel(w, r, currentAuthLevel|AuthTypeOkta2FA)
+	if err != nil {
+		logger.Printf("Auth Cookie NOT found ? %s", err)
+		state.writeFailureResponse(w, r, http.StatusInternalServerError, "Failure when validating VIP token")
+		return
+	}
+	// Now we send to the appropiate place
+	returnAcceptType := getPreferredAcceptType(r)
+	// TODO: The cert backend should depend also on per user preferences.
+	loginResponse := proto.LoginResponse{Message: "success"} //CertAuthBackend: certBackends
+	switch returnAcceptType {
+	case "text/html":
+		loginDestination := getLoginDestination(r)
+		eventNotifier.PublishWebLoginEvent(authUser)
+		http.Redirect(w, r, loginDestination, 302)
+	default:
+		w.WriteHeader(200)
+		json.NewEncoder(w).Encode(loginResponse)
+	}
+	return
+}
+
+///////////////////////////
+const oktaPushStartPath = "/api/v0/oktaPushStart"
+
+func (state *RuntimeState) oktaPushStartHandler(w http.ResponseWriter, r *http.Request) {
+	logger.Printf("top of oktaPushStartHandler")
+	if state.sendFailureToClientIfLocked(w, r) {
+		logger.Printf("foo")
+		return
+	}
+	logger.Printf("oktaPushStartHandler post lock")
+	if !(r.Method == "POST" || r.Method == "GET") {
+		state.writeFailureResponse(w, r, http.StatusMethodNotAllowed, "")
+		return
+	}
+	authUser, _, err := state.checkAuth(w, r, AuthTypeAny)
+	if err != nil {
+		logger.Debugf(1, "%v", err)
+		return
+	}
+	w.(*instrumentedwriter.LoggingWriter).SetUsername(authUser)
+
+	oktaAuth, ok := state.passwordChecker.(*okta.PasswordAuthenticator)
+	if !ok {
+		logger.Debugf(2, "oktaPushStartHandler: password authenticator is not okta is of type %T", oktaAuth)
+		state.writeFailureResponse(w, r, http.StatusInternalServerError, "Apperent Misconfiguration")
+		return
+	}
+	pushResponse, err := oktaAuth.ValidateUserPush(authUser)
+	if err != nil {
+		logger.Println(err)
+		state.writeFailureResponse(w, r, http.StatusInternalServerError, "Failure when validating OKTA push")
+		return
+	}
+	switch pushResponse {
+	case okta.PushResponseWaiting:
+		w.WriteHeader(http.StatusOK)
+		return
+	default:
+		state.writeFailureResponse(w, r, http.StatusPreconditionFailed, "Push already sent")
+		return
+	}
+}
+
+////////////////////////////
+const oktaPollCheckPath = "/api/v0/oktaPollCheck"
+
+func (state *RuntimeState) oktaPollCheckHandler(w http.ResponseWriter, r *http.Request) {
+	if state.sendFailureToClientIfLocked(w, r) {
+		return
+	}
+	if !(r.Method == "POST" || r.Method == "GET") {
```

Comment:
I think this would be more readable:
`if (r.Method != "POST" && r.Method != "GET") {`